src/iterators.rs

```diff
@@ -471,14 +478,32 @@ macro_rules! custom_vec_iter_impl {
                 Ok(self.$data.len())
             }
 
-            fn __getitem__(&self, idx: isize) -> PyResult<$T> {
-                if idx.abs() >= self.$data.len().try_into().unwrap() {
-                    Err(PyIndexError::new_err(format!("Invalid index, {}", idx)))
-                } else if idx < 0 {
-                    let len = self.$data.len();
-                    Ok(self.$data[len - idx.abs() as usize].clone())
-                } else {
-                    Ok(self.$data[idx as usize].clone())
+            fn __getitem__(&self, py: Python, idx: SliceOrInt) -> PyResult<PyObject> {
+                match idx {
+                    SliceOrInt::Slice(slc) => {
+                        let len = self.$data.len().try_into().unwrap();
+                        let indices = slc.indices(len)?;
+                        let start: usize = indices.start.try_into().unwrap();
+                        let stop: usize = indices.stop.try_into().unwrap();
+                        let step: usize = indices.step.try_into().unwrap();
```

Comment:
`unwrap` may panic with negative values. We should either support them or catch the error and return a python exception with a useful error message.

Reply:
I think that might be ok, I don't think that these can actually be negative. I was looking at that c api docs to confirm https://docs.python.org/3/c-api/slice.html but it doesn't actually say. I just don't know how the api could work if any of the values were negative. (well maybe `step` could be if stop < start). I'll do some more testing to see if I can get this to panic with different slice inputs.

Reply:
Yeah I already checked and
```python
import retworkx

graph = retworkx.generators.path_graph(5)
nodes = graph.node_indices()

# panics
sl = nodes[4:1:-1]

# panics
sl = nodes[-1:-3:-1]
```
but if `nodes` was a `list` it'd work.

Reply:
This was a bit trickier to fix than I was hoping, but I think I've got it working now with negative steps with: https://github.com/Qiskit/retworkx/pull/591/commits/fa549b3ff5854aef94de4c90389624041d9bc386